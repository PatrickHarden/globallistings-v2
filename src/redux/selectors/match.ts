import { createMatchSelector, RouterRootState } from 'connected-react-router';
import { PlpPath } from '../../types/plp';
import { PdpPath } from '../../types/pdp';
import RoutePaths from '../../constants/routePaths';
import { createSelector } from 'reselect';

export const plpMatchSelector = createMatchSelector<RouterRootState, PlpPath>(RoutePaths.plp);
export const pdpMatchSelector = createMatchSelector<RouterRootState, PdpPath>(RoutePaths.pdp);

export const pathParamsSelector = createSelector(
  plpMatchSelector,
  pdpMatchSelector,
  (plpMatch, pdpMatch) => ({ ...(plpMatch || {}), ...(pdpMatch || {}) })
);

export const plpUrlPathParamsSelector = createSelector(
  plpMatchSelector,
  match => {
    if (!match) {
      throw new Error('Url path must be defined to use this selector');
    }
    return match.params;
  }
);

export const pdpUrlPathParamsSelector = createSelector(
  pdpMatchSelector,
  match => {
    if (!match) {
      throw new Error('Url path must be defined to use this selector');
    }
    return match.params;
  }
);
