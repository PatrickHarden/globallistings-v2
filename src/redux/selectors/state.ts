import { State } from '../../types/state';
import { theme } from '../../themes/default/theme';
import { createSelector } from 'reselect';
import { FilterType } from '../../types/config';

// State
export const entityMetaSelector = (state: State) => state.entitiesMeta;
export const entitiesSelector = (state: State) => state.entities;
export const uiSelector = (state: State) => state.ui;
export const locationSelector = (state: State) => state.router.location;

// Entities
export const propertiesSelector = createSelector(
  entitiesSelector,
  entities => entities.properties
);

export const geocodeSelector = createSelector(
  entitiesSelector,
  entities => entities.geocode
);

export const propertySearchesSelector = createSelector(
  entitiesSelector,
  entities => entities.propertySearches
);

export const configSelector = createSelector(
  entitiesSelector,
  entities => entities.config
);

export const autocompleteSelector = createSelector(
  entitiesSelector,
  entities => entities.autocomplete
);

export const autocompleteSearchesSelector = createSelector(
  entitiesSelector,
  entities => entities.autocompleteSearches
);

// Meta
export const propertiesMetaSelector = createSelector(
  entityMetaSelector,
  meta => meta.propertySearches
);

export const geocodeMetaSelector = createSelector(
  entityMetaSelector,
  meta => meta.geocode
);

export const configMetaSelector = createSelector(
  entityMetaSelector,
  meta => meta.config
);

export const autocompleteSearchesMetaSelector = createSelector(
  entityMetaSelector,
  meta => meta.autocompleteSearches
);

// Config
export const localeSelector = createSelector(
  configSelector,
  config => config.language
);

export const messagesSelector = createSelector(
  configSelector,
  config => config.i18n
);

export const filtersSelector = createSelector(
  configSelector,
  config => config.filters
);

export const rangeFiltersSelector = createSelector(
  filtersSelector,
  filters => filters.filter(f => f.type === FilterType.range)
);

// Other

// TODO add logic to switch between themes based on config
export const themeSelector = (state: State) => theme;

// TODO figure out where it will be coming from
export const googleMapsApiKeySelector = () => 'AIzaSyCzEw2NaQFixR7sZCLnZ1oGQljW7AxLZ7k';
