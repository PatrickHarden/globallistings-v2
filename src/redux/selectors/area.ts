import { createSelector } from 'reselect';

import { Bounds, LatLng } from '../../types/state';
import { DistanceUnit } from '../../types/config';

import { googleBoundsToBounds } from '../../utils/bounds';
import { haversineRadius } from '../../utils/haversine';
import { paramsSelector } from './config/config';
import { defaultFiltersParamsSelectors } from './config/filters';

import { plpLocationRadiusSelector, plpLocationSelector, locationParamsSelector } from './location';

interface BaseArea {
  readonly bounds: Bounds;
  readonly center: LatLng;
}

export interface RectangleArea extends BaseArea {
  readonly type: 'rectangle';
}

export interface CircleArea extends BaseArea {
  readonly radius: number;
  readonly type: 'circle';
}

export type Area = RectangleArea | CircleArea;

const defaultRadiusSelector = createSelector(
  defaultFiltersParamsSelectors,
  ({ radius }) => {
    if (Array.isArray(radius)) {
      throw new Error('radius can not be an array');
    }

    return radius;
  }
);

const radiusSelector = createSelector(
  defaultRadiusSelector,
  plpLocationRadiusSelector,
  (defaultRadius, radius) => (radius ? radius : defaultRadius)
);

const radiusTypeSelector = createSelector(
  paramsSelector,
  params => (params && params.RadiusType === 'Kilometers' ? DistanceUnit.km : DistanceUnit.miles)
);

export const areaSelector = createSelector(
  radiusSelector,
  plpLocationSelector,
  radiusTypeSelector,
  (radius, location, radiusType): Area => {
    if (!radius || radius === '' || radius === '0') {
      return {
        type: 'rectangle',
        bounds: location.bounds,
        center: location.center,
      };
    }

    const resultRadius = haversineRadius(
      location.bounds,
      location.center,
      Number.parseFloat(radius),
      radiusType
    );

    const boundsCircle = new google.maps.Circle({
      center: location.center,
      radius: resultRadius * 1000,
    });

    return {
      type: 'circle',
      radius: resultRadius,
      bounds: googleBoundsToBounds(boundsCircle.getBounds()),
      center: location.center,
    };
  }
);
