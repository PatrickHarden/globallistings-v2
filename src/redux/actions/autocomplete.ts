import { createStandardAction } from 'typesafe-actions';
import { State } from '../../types/state';
import { autocompleteSelector } from '../selectors/state';
import { updatePath } from './navigation';

export const selectedAutocompletionItem = createStandardAction('SELECT_AUTOCOMPLETION_ITEM')<
  string
>();

export const updateUrlWithSelectedLocation = (placeId: string) => (
  dispatch: Function,
  getState: () => State
) => {
  const place = autocompleteSelector(getState())[placeId];
  if (!!place) {
    dispatch(updatePath('location', place.description));
  }
};
