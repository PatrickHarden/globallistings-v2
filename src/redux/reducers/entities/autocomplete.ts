import { ActionType } from 'typesafe-actions';
import { suggestionsSuccess } from '../../actions/autocomplete-request';
import { Dictionary } from '../../../types/common/dictionary';
import { Autocomplete } from '../../../types/autocomplete';

export type GeocodeActions = ActionType<typeof suggestionsSuccess>;

export default (
  state: Dictionary<Autocomplete> = {},
  action: GeocodeActions
): Dictionary<Autocomplete> => {
  switch (action.type) {
    case 'GET_SUGGESTIONS_SUCCESS':
      return { ...state, ...action.payload.entities.suggestions };
    default:
      return state;
  }
};
