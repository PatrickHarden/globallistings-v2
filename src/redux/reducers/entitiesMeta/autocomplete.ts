import { ActionType } from 'typesafe-actions';
import * as actions from '../../actions/autocomplete-request';
import { handlePendingState, handleErrorState, handleSuccessState } from './shared/meta-reducers';

export type Actions = ActionType<typeof actions>;

export default (state = {}, action: Actions) => {
  switch (action.type) {
    case 'GET_SUGGESTIONS_REQUEST':
      return handlePendingState(state, action);
    case 'GET_SUGGESTIONS_SUCCESS':
      return handleSuccessState(state, action);
    case 'GET_SUGGESTIONS_FAILURE':
      return handleErrorState(state, action);
    default:
      return state;
  }
};
