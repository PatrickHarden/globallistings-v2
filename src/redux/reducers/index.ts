import entities from './entities';
import entitiesMeta from './entitiesMeta';
import ui from './ui';

export default {
  entities,
  entitiesMeta,
  ui,
};
