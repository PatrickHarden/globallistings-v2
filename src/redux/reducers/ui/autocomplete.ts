import { ActionType } from 'typesafe-actions';
import { suggestionsSuccess } from '../../actions/autocomplete-request';
import { selectedAutocompletionItem } from '../../actions/autocomplete';

import { DeepPartial } from 'redux';
import { Autocomplete } from '../../../types/ui';

export type GeocodeActions = ActionType<
  typeof suggestionsSuccess | typeof selectedAutocompletionItem
>;

export default (
  state: DeepPartial<Autocomplete> = {},
  action: GeocodeActions
): DeepPartial<Autocomplete> => {
  switch (action.type) {
    case 'GET_SUGGESTIONS_SUCCESS':
      return { ...state, searchTerm: action.meta.hash };
    case 'SELECT_AUTOCOMPLETION_ITEM':
      return { ...state, selectedSuggestionId: action.payload };
    default:
      return state;
  }
};
