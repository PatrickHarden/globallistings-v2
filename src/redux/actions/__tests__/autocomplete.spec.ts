jest.mock('../../selectors/match');

import { selectedAutocompletionItem, updateUrlWithSelectedLocation } from '../autocomplete';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { State } from '../../../types/state';
import { DeepPartial } from 'redux';
import { CALL_HISTORY_METHOD } from 'connected-react-router';
import { plpUrlPathParamsSelector } from '../../selectors/match';

const middlewares = [thunk];
const mockStore = configureMockStore<State>(middlewares);

((plpUrlPathParamsSelector as unknown) as jest.Mock).mockReturnValue({
  location: 'london',
  usageType: 'office',
  transactionType: 'sale',
});

describe(selectedAutocompletionItem, () => {
  it('should create correct action', () => {
    expect(selectedAutocompletionItem('zpa')).toMatchSnapshot();
  });
});

describe(updateUrlWithSelectedLocation, () => {
  it('should create correct action', () => {
    const state: DeepPartial<State> = {
      router: {
        location: {},
      },
      entities: {
        autocomplete: {
          MOCK_ID: { description: 'zap' },
        },
      },
    };

    const store = mockStore(state as State);

    store.dispatch(updateUrlWithSelectedLocation('MOCK_ID') as any);

    expect(store.getActions()).toEqual([
      {
        payload: { args: ['/office/sale/zap/results'], method: 'replace' },
        type: CALL_HISTORY_METHOD,
      },
    ]);
  });
});
