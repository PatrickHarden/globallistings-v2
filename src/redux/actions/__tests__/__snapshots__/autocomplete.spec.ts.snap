// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create correct action 1`] = `
Object {
  "meta": undefined,
  "payload": "zpa",
  "type": "SELECT_AUTOCOMPLETION_ITEM",
}
`;
